aws-lc-rs/src/kem.rs

```diff
@@ -0,0 +1,530 @@
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+//! Key-Encapsulation Mechanisms (KEMs), including support for Kyber Round 3 Submission.
+//!
+//! # Example
+//!
+//! Note that this example uses the Kyber-512 Round 3 algorithm, but other algorithms can be used
+//! in the exact same way by substituting
+//! `kem::<desired_algorithm_here>` for `kem::KYBER512_R3`.
+//!
+//! ```ignore
+//! use aws_lc_rs::{
+//!     error::Unspecified,
+//!     kem::{Ciphertext, DecapsulationKey, EncapsulationKey},
+//!     unstable::kem::{AlgorithmId, get_algorithm}
+//! };
+//!
+//! let kyber512_r3 = get_algorithm(AlgorithmId::Kyber512_R3).ok_or(Unspecified)?;
+//!
+//! // Alice generates their (private) decapsulation key.
+//! let decapsulation_key = DecapsulationKey::generate(kyber512_r3)?;
+//!
+//! // Alices computes the (public) encapsulation key.
+//! let encapsulation_key = decapsulation_key.encapsulation_key()?;
+//!
+//! // Alice sends the public key bytes to bob through some
+//! // protocol message.
+//! let encapsulation_key_bytes = encapsulation_key.as_ref();
+//!
+//! // Bob constructs the (public) encapsulation key from the key bytes provided by Alice.
+//! let retrieved_encapsulation_key = EncapsulationKey::new(kyber512_r3, encapsulation_key_bytes)?;
+//!
+//! // Bob executes the encapsulation algorithm to to produce their copy of the secret, and associated ciphertext.
+//! let (ciphertext, bob_secret) = retrieved_encapsulation_key.encapsulate()?;
+//!
+//! // Alice recieves ciphertext bytes from bob
+//! let ciphertext_bytes = ciphertext.as_ref();
+//!
+//! // Bob sends Alice the ciphertext computed from the encapsulation algorithm, Alice runs decapsulation to derive their
+//! // copy of the secret.
+//! let alice_secret = decapsulation_key.decapsulate(Ciphertext::from(ciphertext_bytes))?;
+//!
+//! // Alice and Bob have now arrived to the same secret
+//! assert_eq!(alice_secret.as_ref(), bob_secret.as_ref());
+//!
+//! # Ok::<(), aws_lc_rs::error::Unspecified>(())
+//! ```
+use crate::{
+    error::{KeyRejected, Unspecified},
+    ptr::LcPtr,
+    ptr::Pointer,
+};
+use aws_lc::{
+    EVP_PKEY_CTX_kem_set_params, EVP_PKEY_CTX_new, EVP_PKEY_CTX_new_id, EVP_PKEY_decapsulate,
+    EVP_PKEY_encapsulate, EVP_PKEY_get_raw_private_key, EVP_PKEY_get_raw_public_key,
+    EVP_PKEY_kem_new_raw_public_key, EVP_PKEY_kem_new_raw_secret_key, EVP_PKEY_keygen,
+    EVP_PKEY_keygen_init, EVP_PKEY, EVP_PKEY_KEM,
+};
+use std::{borrow::Cow, cmp::Ordering, fmt::Debug, ptr::null_mut};
+use zeroize::Zeroize;
+
+/// An identifier for a KEM algorithm.
+pub trait AlgorithmIdentifier:
+    Copy + Clone + Debug + PartialEq + crate::sealed::Sealed + 'static
+{
+    /// Returns the algorithm's associated AWS-LC nid.
+    fn nid(self) -> i32;
+}
+
+/// A KEM algorithm
+#[derive(PartialEq)]
+pub struct Algorithm<Id = AlgorithmId>
+where
+    Id: AlgorithmIdentifier,
+{
+    pub(crate) id: Id,
+    pub(crate) secret_key_size: usize,
+    pub(crate) public_key_size: usize,
```

Comment:
should these be `encapsulation/decapsulation_key_size` now that we renamed the types?